# BiLSTM

## 1.- Explicación

Un BiLSTM (Bidirectional Long Short-Term Memory) es una variante de las redes neuronales recurrentes (RNN) diseñada para procesar secuencias (texto, audio, series temporales) capturando información tanto del pasado como del “futuro” inmediato de cada elemento en la serie. El BiLSTM amplía esa mirada: mientras un LSTM procesa la secuencia normal, otro procesa la misma secuencia al revés, y luego juntas sus salidas. Así, cada paso tiene un vector de características que incluye señales del pasado y del futuro inmediato.

## 2.- Implementación

### Librerías e instalaciones

In [ ]:
%pip install pandas scikit-learn numpy
%pip install scikit-learn
%pip install numpy
%pip install tensorflow


In [ ]:
import pandas as pd
import numpy as np
import datetime
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler, LabelEncoder
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix, classification_report
from tensorflow.keras.models import Sequential, Model
from tensorflow.keras.layers import Embedding, Bidirectional, LSTM, Attention, Dense, GlobalAveragePooling1D, Input, Dropout, Concatenate
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.utils import to_categorical
import string
import spacy
from datetime import datetime



A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.2.5 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "c:\Users\anaeg\Desktop\LBBYs_CH2\tf_env\Lib\site-packages\ipykernel_launcher.py", line 18, in <module>
    app.launch_new_instance()
  File "c:\Users\anaeg\Desktop\LBBYs_CH2\tf_env\Lib\site-packages\traitlets\config\application.py", line 1075, in launch_instance
    app.start()
  File "c:\Users\anaeg\Desktop\LBBYs_CH2\tf_env\Lib\site-packages\ipykernel\kernelapp.py", line 739, in start
    self.io_loop.start()
 

AttributeError: _ARRAY_API not found


A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.2.5 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "c:\Users\anaeg\Desktop\LBBYs_CH2\tf_env\Lib\site-packages\ipykernel_launcher.py", line 18, in <module>
    app.launch_new_instance()
  File "c:\Users\anaeg\Desktop\LBBYs_CH2\tf_env\Lib\site-packages\traitlets\config\application.py", line 1075, in launch_instance
    app.start()
  File "c:\Users\anaeg\Desktop\LBBYs_CH2\tf_env\Lib\site-packages\ipykernel\kernelapp.py", line 739, in start
    self.io_loop.start()
 

AttributeError: _ARRAY_API not found

ImportError: numpy.core._multiarray_umath failed to import

ImportError: numpy.core.umath failed to import


A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.2.5 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "c:\Users\anaeg\Desktop\LBBYs_CH2\tf_env\Lib\site-packages\ipykernel_launcher.py", line 18, in <module>
    app.launch_new_instance()
  File "c:\Users\anaeg\Desktop\LBBYs_CH2\tf_env\Lib\site-packages\traitlets\config\application.py", line 1075, in launch_instance
    app.start()
  File "c:\Users\anaeg\Desktop\LBBYs_CH2\tf_env\Lib\site-packages\ipykernel\kernelapp.py", line 739, in start
    self.io_loop.start()
 

AttributeError: _ARRAY_API not found

ImportError: numpy.core._multiarray_umath failed to import

ImportError: numpy.core.umath failed to import


A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.2.5 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "c:\Users\anaeg\Desktop\LBBYs_CH2\tf_env\Lib\site-packages\ipykernel_launcher.py", line 18, in <module>
    app.launch_new_instance()
  File "c:\Users\anaeg\Desktop\LBBYs_CH2\tf_env\Lib\site-packages\traitlets\config\application.py", line 1075, in launch_instance
    app.start()
  File "c:\Users\anaeg\Desktop\LBBYs_CH2\tf_env\Lib\site-packages\ipykernel\kernelapp.py", line 739, in start
    self.io_loop.start()
 

AttributeError: _ARRAY_API not found

ImportError: numpy.core._multiarray_umath failed to import

ImportError: numpy.core.umath failed to import

TypeError: Unable to convert function return value to a Python type! The signature was
	() -> handle

In [ ]:
df = pd.read_csv("../../../../data/processed/train_simp_preprocess_v2.csv")
df_test = pd.read_csv("../../../../data/processed/test_simp_preprocess_v2.csv")


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8950 entries, 0 to 8949
Data columns (total 31 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   id                                  8950 non-null   object
 1   label                               8950 non-null   int64 
 2   statement                           8950 non-null   object
 3   subject                             8950 non-null   object
 4   speaker                             8950 non-null   object
 5   speaker_job                         8950 non-null   object
 6   state_info                          8950 non-null   object
 7   party_affiliation                   8950 non-null   object
 8   party_affiliation_uni               8950 non-null   object
 9   party_affiliation_category_map      8950 non-null   object
 10  statement_tokens                    8950 non-null   object
 11  num_tokens                          8950 non-null   int6

### Intento 1: Flujo minimalista con BiLSTM y atención


In [ ]:
print(df.head())

def preprocess_text(text):
    text = text.lower()  
    text = ''.join([char for char in text if char.isalnum() or char.isspace()])  
    return text

df['processed_statement'] = df['statement_tokens_without_stopwords'].apply(preprocess_text)

max_vocab = 10000 
tokenizer = Tokenizer(num_words=max_vocab)
tokenizer.fit_on_texts(df['processed_statement'])
X = tokenizer.texts_to_sequences(df['processed_statement'])

MAX_SEQUENCE_LENGTH = 100  
X = pad_sequences(X, maxlen=MAX_SEQUENCE_LENGTH)

y = df['label'].values

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

input_layer = Input(shape=(MAX_SEQUENCE_LENGTH,))

embedding_layer = Embedding(input_dim=max_vocab, output_dim=100, input_length=MAX_SEQUENCE_LENGTH)(input_layer)

lstm_layer = Bidirectional(LSTM(64, return_sequences=True))(embedding_layer)

attention_layer = Attention()([lstm_layer, lstm_layer])  Aplicar atención entre la secuencia y sí misma

pooling_layer = GlobalAveragePooling1D()(attention_layer)

output_layer = Dense(1, activation='sigmoid')(pooling_layer)

model = Model(inputs=input_layer, outputs=output_layer)

model.compile(optimizer=Adam(), loss='binary_crossentropy', metrics=['accuracy'])

model.summary()

model.fit(X_train, y_train, epochs=3, batch_size=32, validation_data=(X_test, y_test))


            id  label                                          statement  \
0  81f884c64a7      1  china is in the south china sea and (building)...   
1  30c2723a188      0  with the resources it takes to execute just ov...   
2  6936b216e5d      0  the (wisconsin) governor has proposed tax give...   
3  b5cd9195738      1  says her representation of an ex-boyfriend who...   
4  84f8dac7737      0  at protests in wisconsin against proposed coll...   

       subject                     speaker      speaker_job        state_info  \
0        other                donald-trump  president-elect          new york   
1  health-care                  chris-dodd     u.s. senator       connecticut   
2        other               donna-brazile            other  washington, d.c.   
3        other             rebecca-bradley       non-define        non-define   
4        other  republican-party-wisconsin       non-define         wisconsin   

  party_affiliation party_affiliation_uni party_affiliat

c:\Users\anaeg\Desktop\LBBYs_CH2\spacy_env\Lib\site-packages\keras\src\layers\core\embedding.py:90: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer         │ (None, 100)       │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ embedding           │ (None, 100, 100)  │  1,000,000 │ input_layer[0][0] │
│ (Embedding)         │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ bidirectional       │ (None, 100, 128)  │     84,480 │ embedding[0][0]   │
│ (Bidirectional)     │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ attention           │ (None, 100, 128)  │          0 │ bidirectional[0]… │
│ (Attention)         │                   │            │ bidirectional[0]… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ global_average_poo… │ (None, 128)       │          0 │ attention[0][0]   │
│ (GlobalAveragePool… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense (Dense)       │ (None, 1)         │        129 │ global_average_p… │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 1,084,609 (4.14 MB)

 Trainable params: 1,084,609 (4.14 MB)

 Non-trainable params: 0 (0.00 B)

Epoch 1/3
224/224 ━━━━━━━━━━━━━━━━━━━━ 7s 22ms/step - accuracy: 0.6303 - loss: 0.6566 - val_accuracy: 0.6587 - val_loss: 0.6471
Epoch 2/3
224/224 ━━━━━━━━━━━━━━━━━━━━ 5s 21ms/step - accuracy: 0.6454 - loss: 0.6504 - val_accuracy: 0.6587 - val_loss: 0.6423
Epoch 3/3
224/224 ━━━━━━━━━━━━━━━━━━━━ 5s 21ms/step - accuracy: 0.6468 - loss: 0.6406 - val_accuracy: 0.6559 - val_loss: 0.6321


In [ ]:
import datetime
from sklearn.metrics import confusion_matrix, classification_report
import pandas as pd
import numpy as np
from tensorflow.keras.preprocessing.sequence import pad_sequences


df_test['processed_statement'] = df_test['statement_tokens_without_stopwords'].apply(preprocess_text)

X_test_df = tokenizer.texts_to_sequences(df_test['processed_statement'])

X_test_df = pad_sequences(X_test_df, maxlen=MAX_SEQUENCE_LENGTH)

y_pred = model.predict(X_test_df)
y_pred = (y_pred > 0.5).astype(int)  

submission = pd.DataFrame({
    'id': df_test['id'],
    'label': y_pred.flatten()  
})

filename = f"submission_v2_{datetime.datetime.now().strftime('%Y%m%d_%H%M')}.csv"
submission.to_csv(filename, columns=['id', 'label'], index=False)

print(f" Submission generada correctamente: '{filename}'")
print(submission.head())


ValueError: Layer "functional_1" expects 2 input(s), but it received 1 input tensors. Inputs received: [<tf.Tensor 'data:0' shape=(32, 100) dtype=int32>]

### Intento 2:  Preprocesado avanzado con spaCy y metadatos categóricos

In [14]:
from datetime import datetime

In [ ]:
nlp = spacy.load("en_core_web_sm")


In [ ]:
def preprocess_text_advanced(text):
    text = text.lower()
    text = text.translate(str.maketrans('', '', string.punctuation))
    doc = nlp(text)
    tokens = [token.lemma_ for token in doc if not token.is_stop and not token.is_punct]
    return " ".join(tokens)

df['processed_statement'] = df['statement'].apply(preprocess_text_advanced)
df_test['processed_statement'] = df_test['statement'].apply(preprocess_text_advanced)

max_vocab = 10000 
tokenizer = Tokenizer(num_words=max_vocab)
tokenizer.fit_on_texts(df['processed_statement'])

X_train = tokenizer.texts_to_sequences(df['processed_statement'])
X_test = tokenizer.texts_to_sequences(df_test['processed_statement'])

MAX_SEQUENCE_LENGTH = 100  
X_train = pad_sequences(X_train, maxlen=MAX_SEQUENCE_LENGTH)
X_test = pad_sequences(X_test, maxlen=MAX_SEQUENCE_LENGTH)

subject_encoded = pd.get_dummies(df['processed_subject'])
speaker_encoded = pd.get_dummies(df['speaker'])
party_affiliation_encoded = pd.get_dummies(df['party_affiliation'])

X_additional_features = pd.concat([subject_encoded, speaker_encoded, party_affiliation_encoded], axis=1)
y = df['label'].values

from sklearn.model_selection import train_test_split
X_train, X_val, y_train, y_val = train_test_split(X_train, y, test_size=0.2, random_state=42)

input_text = Input(shape=(MAX_SEQUENCE_LENGTH,))
input_additional = Input(shape=(X_additional_features.shape[1],))  

embedding_layer = Embedding(input_dim=max_vocab, output_dim=100, input_length=MAX_SEQUENCE_LENGTH)(input_text)

lstm_layer = Bidirectional(LSTM(128, return_sequences=True))(embedding_layer)
dropout_layer = Dropout(0.5)(lstm_layer)

attention_layer = Attention()([dropout_layer, dropout_layer])

pooling_layer = GlobalAveragePooling1D()(attention_layer)

merged = Concatenate()([pooling_layer, input_additional])

output_layer = Dense(1, activation='sigmoid')(merged)

model = Model(inputs=[input_text, input_additional], outputs=output_layer)

model.compile(optimizer=Adam(), loss='binary_crossentropy', metrics=['accuracy'])

model.summary()

model.fit([X_train, X_additional_features.iloc[:len(X_train)]], y_train, epochs=5, batch_size=32, validation_data=([X_val, X_additional_features.iloc[:len(X_val)]], y_val))

y_pred_prob = model.predict([X_test, X_additional_features.iloc[:len(X_test)]]) 
y_pred = (y_pred_prob > 0.5).astype(int)  

submission = pd.DataFrame({
    'id': df_test['id'],
    'label': y_pred.flatten()  
})


filename = f"biLSTM_submission_v2_{datetime.now().strftime('%Y%m%d')}.csv"
submission.to_csv(filename, columns=['id', 'label'], index=False)

print(f" Submission generada correctamente: '{filename}'")
print(submission.head())

c:\Users\anaeg\Desktop\LBBYs_CH2\spacy_env\Lib\site-packages\keras\src\layers\core\embedding.py:90: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


Model: "functional_3"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer_5       │ (None, 100)       │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ embedding_3         │ (None, 100, 100)  │  1,000,000 │ input_layer_5[0]… │
│ (Embedding)         │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ bidirectional_3     │ (None, 100, 256)  │    234,496 │ embedding_3[0][0] │
│ (Bidirectional)     │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_2 (Dropout) │ (None, 100, 256)  │          0 │ bidirectional_3[… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ attention_3         │ (None, 100, 256)  │          0 │ dropout_2[0][0],  │
│ (Attention)         │                   │            │ dropout_2[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ global_average_poo… │ (None, 256)       │          0 │ attention_3[0][0] │
│ (GlobalAveragePool… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ input_layer_6       │ (None, 2746)      │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ concatenate_2       │ (None, 3002)      │          0 │ global_average_p… │
│ (Concatenate)       │                   │            │ input_layer_6[0]… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_3 (Dense)     │ (None, 1)         │      3,003 │ concatenate_2[0]… │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 1,237,499 (4.72 MB)

 Trainable params: 1,237,499 (4.72 MB)

 Non-trainable params: 0 (0.00 B)

Epoch 1/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 11s 36ms/step - accuracy: 0.6426 - loss: 0.6583 - val_accuracy: 0.6587 - val_loss: 0.6449
Epoch 2/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 7s 33ms/step - accuracy: 0.6393 - loss: 0.6520 - val_accuracy: 0.6587 - val_loss: 0.6561
Epoch 3/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 7s 33ms/step - accuracy: 0.6418 - loss: 0.6507 - val_accuracy: 0.6581 - val_loss: 0.6487
Epoch 4/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 8s 36ms/step - accuracy: 0.6797 - loss: 0.6042 - val_accuracy: 0.6028 - val_loss: 0.6681
Epoch 5/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 8s 37ms/step - accuracy: 0.7121 - loss: 0.5698 - val_accuracy: 0.6201 - val_loss: 0.6670
120/120 ━━━━━━━━━━━━━━━━━━━━ 2s 14ms/step
✅ Submission generada correctamente: 'biLSTM_submission_v2_20250515.csv'
            id  label
0  dc32e5ffa8b      1
1  aa49bb41cab      1
2  dddc8d12ac1      1
3  bcfe8f51667      1
4  eedbbaff5ab      1


### Intento 3: LSTM ampliado con pesos de clase balanceados

In [ ]:
import pandas as pd
import numpy as np
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.layers import Dense, Embedding, Bidirectional, LSTM, GRU, Input, Dropout, GlobalAveragePooling1D, Attention, Concatenate
from tensorflow.keras.models import Model
from tensorflow.keras.optimizers import Adam
from sklearn.metrics import confusion_matrix, classification_report
from tensorflow.keras.callbacks import ReduceLROnPlateau
from datetime import datetime
import string
import spacy
from sklearn.utils.class_weight import compute_class_weight

nlp = spacy.load("en_core_web_sm")

def preprocess_text_advanced(text):
    text = text.lower()
    text = text.translate(str.maketrans('', '', string.punctuation))
    doc = nlp(text)
    tokens = [token.lemma_ for token in doc if not token.is_stop and not token.is_punct]
    return " ".join(tokens)

df['processed_statement'] = df['statement'].apply(preprocess_text_advanced)
df_test['processed_statement'] = df_test['statement'].apply(preprocess_text_advanced)

max_vocab = 10000  
tokenizer = Tokenizer(num_words=max_vocab)
tokenizer.fit_on_texts(df['processed_statement'])

X_train = tokenizer.texts_to_sequences(df['processed_statement'])
X_test = tokenizer.texts_to_sequences(df_test['processed_statement'])

MAX_SEQUENCE_LENGTH = 100  
X_train = pad_sequences(X_train, maxlen=MAX_SEQUENCE_LENGTH)
X_test = pad_sequences(X_test, maxlen=MAX_SEQUENCE_LENGTH)

subject_encoded = pd.get_dummies(df['processed_subject'])
speaker_encoded = pd.get_dummies(df['speaker'])
party_affiliation_encoded = pd.get_dummies(df['party_affiliation'])

X_additional_features = pd.concat([subject_encoded, speaker_encoded, party_affiliation_encoded], axis=1)
y = df['label'].values

from sklearn.model_selection import train_test_split
X_train, X_val, y_train, y_val = train_test_split(X_train, y, test_size=0.2, random_state=42)

class_weights = compute_class_weight('balanced', classes=np.unique(y_train), y=y_train)

input_text = Input(shape=(MAX_SEQUENCE_LENGTH,))
input_additional = Input(shape=(X_additional_features.shape[1],)) 

embedding_layer = Embedding(input_dim=max_vocab, output_dim=100, input_length=MAX_SEQUENCE_LENGTH)(input_text)

lstm_layer = Bidirectional(LSTM(256, return_sequences=True))(embedding_layer)  
dropout_layer = Dropout(0.5)(lstm_layer)

attention_layer = Attention()([dropout_layer, dropout_layer])
pooling_layer = GlobalAveragePooling1D()(attention_layer)
merged = Concatenate()([pooling_layer, input_additional])
output_layer = Dense(1, activation='sigmoid')(merged)
model = Model(inputs=[input_text, input_additional], outputs=output_layer)
reduce_lr = ReduceLROnPlateau(monitor='val_loss', factor=0.5, patience=2, min_lr=1e-6)

model.compile(optimizer=Adam(), loss='binary_crossentropy', metrics=['accuracy'])
model.summary()
model.fit([X_train, X_additional_features.iloc[:len(X_train)]], y_train, epochs=5, batch_size=32, 
          validation_data=([X_val, X_additional_features.iloc[:len(X_val)]], y_val), 
          class_weight={0: class_weights[0], 1: class_weights[1]},
          callbacks=[reduce_lr])

y_pred_prob = model.predict([X_test, X_additional_features.iloc[:len(X_test)]])  
y_pred = (y_pred_prob > 0.5).astype(int)  
submission = pd.DataFrame({
    'id': df_test['id'],
    'label': y_pred.flatten()  
})

filename = f"biLSTM_submission_v2_{datetime.now().strftime('%Y%m%d_%H%M')}.csv"
submission.to_csv(filename, columns=['id', 'label'], index=False)

print(f" Submission generada correctamente: '{filename}'")
print(submission.head())


c:\Users\anaeg\Desktop\LBBYs_CH2\spacy_env\Lib\site-packages\keras\src\layers\core\embedding.py:90: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


Model: "functional_4"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer_7       │ (None, 100)       │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ embedding_4         │ (None, 100, 100)  │  1,000,000 │ input_layer_7[0]… │
│ (Embedding)         │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ bidirectional_4     │ (None, 100, 512)  │    731,136 │ embedding_4[0][0] │
│ (Bidirectional)     │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_3 (Dropout) │ (None, 100, 512)  │          0 │ bidirectional_4[… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ attention_4         │ (None, 100, 512)  │          0 │ dropout_3[0][0],  │
│ (Attention)         │                   │            │ dropout_3[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ global_average_poo… │ (None, 512)       │          0 │ attention_4[0][0] │
│ (GlobalAveragePool… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ input_layer_8       │ (None, 2746)      │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ concatenate_3       │ (None, 3258)      │          0 │ global_average_p… │
│ (Concatenate)       │                   │            │ input_layer_8[0]… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_4 (Dense)     │ (None, 1)         │      3,259 │ concatenate_3[0]… │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 1,734,395 (6.62 MB)

 Trainable params: 1,734,395 (6.62 MB)

 Non-trainable params: 0 (0.00 B)

Epoch 1/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 52s 223ms/step - accuracy: 0.5222 - loss: 0.6913 - val_accuracy: 0.5391 - val_loss: 0.6916 - learning_rate: 0.0010
Epoch 2/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 41s 184ms/step - accuracy: 0.5559 - loss: 0.6898 - val_accuracy: 0.4296 - val_loss: 0.6967 - learning_rate: 0.0010
Epoch 3/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 40s 178ms/step - accuracy: 0.5956 - loss: 0.6863 - val_accuracy: 0.4989 - val_loss: 0.6931 - learning_rate: 0.0010
Epoch 4/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 42s 190ms/step - accuracy: 0.6706 - loss: 0.6716 - val_accuracy: 0.6022 - val_loss: 0.7142 - learning_rate: 5.0000e-04
Epoch 5/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 41s 183ms/step - accuracy: 0.6908 - loss: 0.6271 - val_accuracy: 0.6095 - val_loss: 0.7016 - learning_rate: 5.0000e-04
120/120 ━━━━━━━━━━━━━━━━━━━━ 5s 39ms/step
✅ Submission generada correctamente: 'biLSTM_submission_v2_20250515_1116.csv'
            id  label
0  dc32e5ffa8b      0
1  aa49bb41cab      1
2  dddc8d12ac1      1
3  bcfe8f51667

### Intento 4: Variaciones de vocabulario y codificación de metacaracterísticas

In [ ]:
import pandas as pd
import numpy as np
import string
import spacy
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.layers import Dense, Embedding, Bidirectional, LSTM, Input, Dropout, GlobalAveragePooling1D, Attention, Concatenate
from tensorflow.keras.models import Model
from tensorflow.keras.optimizers import Adam
from sklearn.metrics import classification_report
from sklearn.model_selection import train_test_split
from sklearn.utils.class_weight import compute_class_weight
from tensorflow.keras.callbacks import ReduceLROnPlateau
from datetime import datetime


nlp = spacy.load("en_core_web_sm")
max_vocab = 10000 
tokenizer = Tokenizer(num_words=max_vocab)
tokenizer.fit_on_texts(df['processed_statement'])

X_train = tokenizer.texts_to_sequences(df['processed_statement'])
X_test = tokenizer.texts_to_sequences(df_test['processed_statement'])

MAX_SEQUENCE_LENGTH = 100  
X_train = pad_sequences(X_train, maxlen=MAX_SEQUENCE_LENGTH)
X_test = pad_sequences(X_test, maxlen=MAX_SEQUENCE_LENGTH)
subject_encoded = pd.get_dummies(df['processed_subject'])
speaker_encoded = pd.get_dummies(df['speaker'])
party_affiliation_encoded = pd.get_dummies(df['party_affiliation'])
X_additional_features = pd.concat([subject_encoded, speaker_encoded, party_affiliation_encoded], axis=1)

y = df['label'].values

X_train, X_val, y_train, y_val = train_test_split(X_train, y, test_size=0.2, random_state=42)
class_weights = compute_class_weight('balanced', classes=np.unique(y_train), y=y_train)

input_text = Input(shape=(MAX_SEQUENCE_LENGTH,))
input_additional = Input(shape=(X_additional_features.shape[1],)) 

embedding_layer = Embedding(input_dim=max_vocab, output_dim=100, input_length=MAX_SEQUENCE_LENGTH)(input_text)

lstm_layer = Bidirectional(LSTM(256, return_sequences=True))(embedding_layer)

dropout_layer = Dropout(0.5)(lstm_layer)

attention_layer = Attention()([dropout_layer, dropout_layer])

pooling_layer = GlobalAveragePooling1D()(attention_layer)

merged = Concatenate()([pooling_layer, input_additional])

output_layer = Dense(1, activation='sigmoid')(merged)

model = Model(inputs=[input_text, input_additional], outputs=output_layer)

reduce_lr = ReduceLROnPlateau(monitor='val_loss', factor=0.5, patience=2, min_lr=1e-6)

model.compile(optimizer=Adam(), loss='binary_crossentropy', metrics=['accuracy'])

model.summary()
model.fit([X_train, X_additional_features.iloc[:len(X_train)]], y_train, epochs=5, batch_size=32,
          validation_data=([X_val, X_additional_features.iloc[:len(X_val)]], y_val),
          class_weight={0: class_weights[0], 1: class_weights[1]},
          callbacks=[reduce_lr])
y_pred_prob = model.predict([X_test, X_additional_features.iloc[:len(X_test)]])
y_pred = (y_pred_prob > 0.5).astype(int)  

submission = pd.DataFrame({
    'id': df_test['id'],
    'label': y_pred.flatten()  
})

filename = f"biLSTM_submission_v2_{datetime.now().strftime('%Y%m%d_%H%M')}.csv"
submission.to_csv(filename, columns=['id', 'label'], index=False)

print(f" Submission generada correctamente: '{filename}'")
print(submission.head())


c:\Users\anaeg\Desktop\LBBYs_CH2\spacy_env\Lib\site-packages\keras\src\layers\core\embedding.py:90: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


Model: "functional_5"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer_9       │ (None, 100)       │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ embedding_5         │ (None, 100, 100)  │  1,000,000 │ input_layer_9[0]… │
│ (Embedding)         │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ bidirectional_5     │ (None, 100, 512)  │    731,136 │ embedding_5[0][0] │
│ (Bidirectional)     │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_4 (Dropout) │ (None, 100, 512)  │          0 │ bidirectional_5[… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ attention_5         │ (None, 100, 512)  │          0 │ dropout_4[0][0],  │
│ (Attention)         │                   │            │ dropout_4[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ global_average_poo… │ (None, 512)       │          0 │ attention_5[0][0] │
│ (GlobalAveragePool… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ input_layer_10      │ (None, 2746)      │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ concatenate_4       │ (None, 3258)      │          0 │ global_average_p… │
│ (Concatenate)       │                   │            │ input_layer_10[0… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_5 (Dense)     │ (None, 1)         │      3,259 │ concatenate_4[0]… │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 1,734,395 (6.62 MB)

 Trainable params: 1,734,395 (6.62 MB)

 Non-trainable params: 0 (0.00 B)

Epoch 1/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 42s 175ms/step - accuracy: 0.4916 - loss: 0.6903 - val_accuracy: 0.4810 - val_loss: 0.6945 - learning_rate: 0.0010
Epoch 2/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 35s 157ms/step - accuracy: 0.5292 - loss: 0.6925 - val_accuracy: 0.5112 - val_loss: 0.6933 - learning_rate: 0.0010
Epoch 3/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 34s 152ms/step - accuracy: 0.5543 - loss: 0.6898 - val_accuracy: 0.6196 - val_loss: 0.6827 - learning_rate: 0.0010
Epoch 4/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 35s 155ms/step - accuracy: 0.6189 - loss: 0.6756 - val_accuracy: 0.5737 - val_loss: 0.6799 - learning_rate: 0.0010
Epoch 5/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 35s 157ms/step - accuracy: 0.6652 - loss: 0.6292 - val_accuracy: 0.5620 - val_loss: 0.6930 - learning_rate: 0.0010
120/120 ━━━━━━━━━━━━━━━━━━━━ 4s 32ms/step
✅ Submission generada correctamente: 'biLSTM_submission_v2_20250515_1119.csv'
            id  label
0  dc32e5ffa8b      0
1  aa49bb41cab      1
2  dddc8d12ac1      1
3  bcfe8f51667      1


### Intento 5: Variaciones de vocabulario y codificación de metacaracterísticas

In [ ]:
import pandas as pd
import numpy as np
import string
import spacy
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.layers import Dense, Embedding, Bidirectional, LSTM, Input, Dropout, GlobalAveragePooling1D, Attention, Concatenate
from tensorflow.keras.models import Model
from tensorflow.keras.optimizers import Adam
from sklearn.metrics import classification_report
from sklearn.model_selection import train_test_split
from sklearn.utils.class_weight import compute_class_weight
from tensorflow.keras.callbacks import ReduceLROnPlateau
from datetime import datetime





max_vocab = 10000  
tokenizer = Tokenizer(num_words=max_vocab)
tokenizer.fit_on_texts(df['processed_statement'])


X_train = tokenizer.texts_to_sequences(df['processed_statement'])
X_test = tokenizer.texts_to_sequences(df_test['processed_statement'])


MAX_SEQUENCE_LENGTH = 100  
X_train = pad_sequences(X_train, maxlen=MAX_SEQUENCE_LENGTH)
X_test = pad_sequences(X_test, maxlen=MAX_SEQUENCE_LENGTH)

subject_encoded = pd.get_dummies(df['processed_subject'])
speaker_encoded = pd.get_dummies(df['speaker_job'])
party_affiliation_encoded = pd.get_dummies(df['party_affiliation_uni'])

X_additional_features = pd.concat([subject_encoded, speaker_encoded, party_affiliation_encoded], axis=1)

y = df['label'].values

X_train, X_val, y_train, y_val = train_test_split(X_train, y, test_size=0.2, random_state=42)

class_weights = compute_class_weight('balanced', classes=np.unique(y_train), y=y_train)

input_text = Input(shape=(MAX_SEQUENCE_LENGTH,))
input_additional = Input(shape=(X_additional_features.shape[1],)) 

embedding_layer = Embedding(input_dim=max_vocab, output_dim=100, input_length=MAX_SEQUENCE_LENGTH)(input_text)

lstm_layer = Bidirectional(LSTM(256, return_sequences=True))(embedding_layer)

dropout_layer = Dropout(0.5)(lstm_layer)

attention_layer = Attention()([dropout_layer, dropout_layer])
pooling_layer = GlobalAveragePooling1D()(attention_layer)

merged = Concatenate()([pooling_layer, input_additional])

output_layer = Dense(1, activation='sigmoid')(merged)

model = Model(inputs=[input_text, input_additional], outputs=output_layer)

reduce_lr = ReduceLROnPlateau(monitor='val_loss', factor=0.5, patience=2, min_lr=1e-6)

model.compile(optimizer=Adam(), loss='binary_crossentropy', metrics=['accuracy'])

model.summary()

model.fit([X_train, X_additional_features.iloc[:len(X_train)]], y_train, epochs=5, batch_size=32,
          validation_data=([X_val, X_additional_features.iloc[:len(X_val)]], y_val),
          class_weight={0: class_weights[0], 1: class_weights[1]},
          callbacks=[reduce_lr])

y_pred_prob = model.predict([X_test, X_additional_features.iloc[:len(X_test)]])
y_pred = (y_pred_prob > 0.5).astype(int)  

submission = pd.DataFrame({
    'id': df_test['id'],
    'label': y_pred.flatten()  
})


filename = f"biLSTM_submission_v2_{datetime.now().strftime('%Y%m%d_%H%M')}.csv"
submission.to_csv(filename, columns=['id', 'label'], index=False)

print(f" Submission generada correctamente: '{filename}'")
print(submission.head())


c:\Users\anaeg\Desktop\LBBYs_CH2\spacy_env\Lib\site-packages\keras\src\layers\core\embedding.py:90: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


Model: "functional_6"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer_11      │ (None, 100)       │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ embedding_6         │ (None, 100, 100)  │  1,000,000 │ input_layer_11[0… │
│ (Embedding)         │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ bidirectional_6     │ (None, 100, 512)  │    731,136 │ embedding_6[0][0] │
│ (Bidirectional)     │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_5 (Dropout) │ (None, 100, 512)  │          0 │ bidirectional_6[… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ attention_6         │ (None, 100, 512)  │          0 │ dropout_5[0][0],  │
│ (Attention)         │                   │            │ dropout_5[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ global_average_poo… │ (None, 512)       │          0 │ attention_6[0][0] │
│ (GlobalAveragePool… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ input_layer_12      │ (None, 139)       │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ concatenate_5       │ (None, 651)       │          0 │ global_average_p… │
│ (Concatenate)       │                   │            │ input_layer_12[0… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_6 (Dense)     │ (None, 1)         │        652 │ concatenate_5[0]… │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 1,731,788 (6.61 MB)

 Trainable params: 1,731,788 (6.61 MB)

 Non-trainable params: 0 (0.00 B)

Epoch 1/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 56s 236ms/step - accuracy: 0.5083 - loss: 0.6941 - val_accuracy: 0.5609 - val_loss: 0.6894 - learning_rate: 0.0010
Epoch 2/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 45s 201ms/step - accuracy: 0.5140 - loss: 0.6924 - val_accuracy: 0.3413 - val_loss: 0.7423 - learning_rate: 0.0010
Epoch 3/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 45s 199ms/step - accuracy: 0.4962 - loss: 0.6939 - val_accuracy: 0.4475 - val_loss: 0.6900 - learning_rate: 0.0010
Epoch 4/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 46s 207ms/step - accuracy: 0.6776 - loss: 0.6777 - val_accuracy: 0.6307 - val_loss: 0.6768 - learning_rate: 5.0000e-04
Epoch 5/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 49s 217ms/step - accuracy: 0.6984 - loss: 0.6212 - val_accuracy: 0.6028 - val_loss: 0.7069 - learning_rate: 5.0000e-04
120/120 ━━━━━━━━━━━━━━━━━━━━ 6s 47ms/step
✅ Submission generada correctamente: 'biLSTM_submission_v2_20250515_1123.csv'
            id  label
0  dc32e5ffa8b      0
1  aa49bb41cab      0
2  dddc8d12ac1      1
3  bcfe8f51667

### Intento 6: BiLSTM de 256 unidades con atención y metacaracterísticas adicionales

In [ ]:
import pandas as pd
import numpy as np
import string
import spacy
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.layers import Dense, Embedding, Bidirectional, LSTM, Input, Dropout, GlobalAveragePooling1D, Attention, Concatenate
from tensorflow.keras.models import Model
from tensorflow.keras.optimizers import Adam
from sklearn.metrics import classification_report
from sklearn.model_selection import train_test_split
from sklearn.utils.class_weight import compute_class_weight
from tensorflow.keras.callbacks import ReduceLROnPlateau
from datetime import datetime





max_vocab = 10000  
tokenizer = Tokenizer(num_words=max_vocab)
tokenizer.fit_on_texts(df['processed_statement'])


X_train = tokenizer.texts_to_sequences(df['processed_statement'])
X_test = tokenizer.texts_to_sequences(df_test['processed_statement'])


MAX_SEQUENCE_LENGTH = 100  
X_train = pad_sequences(X_train, maxlen=MAX_SEQUENCE_LENGTH)
X_test = pad_sequences(X_test, maxlen=MAX_SEQUENCE_LENGTH)


subject_encoded = pd.get_dummies(df['processed_subject'])
speaker_encoded = pd.get_dummies(df['speaker_job'])
party_affiliation_encoded = pd.get_dummies(df['party_affiliation_category_map'])


X_additional_features = pd.concat([subject_encoded, speaker_encoded, party_affiliation_encoded], axis=1)


y = df['label'].values


X_train, X_val, y_train, y_val = train_test_split(X_train, y, test_size=0.2, random_state=42)


class_weights = compute_class_weight('balanced', classes=np.unique(y_train), y=y_train)


input_text = Input(shape=(MAX_SEQUENCE_LENGTH,))
input_additional = Input(shape=(X_additional_features.shape[1],))  


embedding_layer = Embedding(input_dim=max_vocab, output_dim=100, input_length=MAX_SEQUENCE_LENGTH)(input_text)


lstm_layer = Bidirectional(LSTM(256, return_sequences=True))(embedding_layer)


dropout_layer = Dropout(0.5)(lstm_layer)


attention_layer = Attention()([dropout_layer, dropout_layer])


pooling_layer = GlobalAveragePooling1D()(attention_layer)


merged = Concatenate()([pooling_layer, input_additional])


output_layer = Dense(1, activation='sigmoid')(merged)


model = Model(inputs=[input_text, input_additional], outputs=output_layer)


reduce_lr = ReduceLROnPlateau(monitor='val_loss', factor=0.5, patience=2, min_lr=1e-6)


model.compile(optimizer=Adam(), loss='binary_crossentropy', metrics=['accuracy'])


model.summary()


model.fit([X_train, X_additional_features.iloc[:len(X_train)]], y_train, epochs=5, batch_size=32,
          validation_data=([X_val, X_additional_features.iloc[:len(X_val)]], y_val),
          class_weight={0: class_weights[0], 1: class_weights[1]},
          callbacks=[reduce_lr])


y_pred_prob = model.predict([X_test, X_additional_features.iloc[:len(X_test)]])
y_pred = (y_pred_prob > 0.5).astype(int)  


submission = pd.DataFrame({
    'id': df_test['id'],
    'label': y_pred.flatten()  
})


filename = f"biLSTM_submission_v2_{datetime.now().strftime('%Y%m%d_%H%M')}.csv"
submission.to_csv(filename, columns=['id', 'label'], index=False)


print(f" Submission generada correctamente: '{filename}'")
print(submission.head())


c:\Users\anaeg\Desktop\LBBYs_CH2\spacy_env\Lib\site-packages\keras\src\layers\core\embedding.py:90: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


Model: "functional_7"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer_13      │ (None, 100)       │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ embedding_7         │ (None, 100, 100)  │  1,000,000 │ input_layer_13[0… │
│ (Embedding)         │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ bidirectional_7     │ (None, 100, 512)  │    731,136 │ embedding_7[0][0] │
│ (Bidirectional)     │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_6 (Dropout) │ (None, 100, 512)  │          0 │ bidirectional_7[… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ attention_7         │ (None, 100, 512)  │          0 │ dropout_6[0][0],  │
│ (Attention)         │                   │            │ dropout_6[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ global_average_poo… │ (None, 512)       │          0 │ attention_7[0][0] │
│ (GlobalAveragePool… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ input_layer_14      │ (None, 131)       │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ concatenate_6       │ (None, 643)       │          0 │ global_average_p… │
│ (Concatenate)       │                   │            │ input_layer_14[0… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_7 (Dense)     │ (None, 1)         │        644 │ concatenate_6[0]… │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 1,731,780 (6.61 MB)

 Trainable params: 1,731,780 (6.61 MB)

 Non-trainable params: 0 (0.00 B)

Epoch 1/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 36s 149ms/step - accuracy: 0.4697 - loss: 0.6991 - val_accuracy: 0.5307 - val_loss: 0.6905 - learning_rate: 0.0010
Epoch 2/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 27s 121ms/step - accuracy: 0.5108 - loss: 0.6924 - val_accuracy: 0.4246 - val_loss: 0.6979 - learning_rate: 0.0010
Epoch 3/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 26s 114ms/step - accuracy: 0.4881 - loss: 0.6928 - val_accuracy: 0.6480 - val_loss: 0.6801 - learning_rate: 0.0010
Epoch 4/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 24s 109ms/step - accuracy: 0.5728 - loss: 0.6879 - val_accuracy: 0.4140 - val_loss: 0.7140 - learning_rate: 0.0010
Epoch 5/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 26s 115ms/step - accuracy: 0.6010 - loss: 0.6639 - val_accuracy: 0.5587 - val_loss: 0.6906 - learning_rate: 0.0010
120/120 ━━━━━━━━━━━━━━━━━━━━ 3s 25ms/step
✅ Submission generada correctamente: 'biLSTM_submission_v2_20250515_1126.csv'
            id  label
0  dc32e5ffa8b      0
1  aa49bb41cab      1
2  dddc8d12ac1      1
3  bcfe8f51667      1


In [ ]:
import pandas as pd
import numpy as np
import string
import spacy
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.layers import Dense, Embedding, Bidirectional, LSTM, Input, Dropout, GlobalAveragePooling1D, Attention, Concatenate
from tensorflow.keras.models import Model
from tensorflow.keras.optimizers import Adam
from sklearn.metrics import classification_report
from sklearn.model_selection import train_test_split
from sklearn.utils.class_weight import compute_class_weight
from tensorflow.keras.callbacks import ReduceLROnPlateau
from datetime import datetime




max_vocab = 20000  
tokenizer = Tokenizer(num_words=max_vocab)
tokenizer.fit_on_texts(df['processed_statement'])


X_train = tokenizer.texts_to_sequences(df['processed_statement'])
X_test = tokenizer.texts_to_sequences(df_test['processed_statement'])


MAX_SEQUENCE_LENGTH = 100  
X_train = pad_sequences(X_train, maxlen=MAX_SEQUENCE_LENGTH)
X_test = pad_sequences(X_test, maxlen=MAX_SEQUENCE_LENGTH)


subject_encoded = pd.get_dummies(df['processed_subject'])
speaker_encoded = pd.get_dummies(df['speaker_job'])
party_affiliation_encoded = pd.get_dummies(df['party_affiliation_uni'])


X_additional_features = pd.concat([subject_encoded, speaker_encoded, party_affiliation_encoded], axis=1)


y = df['label'].values


X_train, X_val, y_train, y_val = train_test_split(X_train, y, test_size=0.2, random_state=42)


class_weights = compute_class_weight('balanced', classes=np.unique(y_train), y=y_train)


input_text = Input(shape=(MAX_SEQUENCE_LENGTH,))
input_additional = Input(shape=(X_additional_features.shape[1],))  


embedding_layer = Embedding(input_dim=max_vocab, output_dim=100, input_length=MAX_SEQUENCE_LENGTH)(input_text)


lstm_layer = Bidirectional(LSTM(256, return_sequences=True))(embedding_layer)


dropout_layer = Dropout(0.5)(lstm_layer)


attention_layer = Attention()([dropout_layer, dropout_layer])


pooling_layer = GlobalAveragePooling1D()(attention_layer)


merged = Concatenate()([pooling_layer, input_additional])


output_layer = Dense(1, activation='sigmoid')(merged)


model = Model(inputs=[input_text, input_additional], outputs=output_layer)


reduce_lr = ReduceLROnPlateau(monitor='val_loss', factor=0.5, patience=2, min_lr=1e-6)


model.compile(optimizer=Adam(), loss='binary_crossentropy', metrics=['accuracy'])


model.summary()


model.fit([X_train, X_additional_features.iloc[:len(X_train)]], y_train, epochs=5, batch_size=32,
          validation_data=([X_val, X_additional_features.iloc[:len(X_val)]], y_val),
          class_weight={0: class_weights[0], 1: class_weights[1]},
          callbacks=[reduce_lr])


y_pred_prob = model.predict([X_test, X_additional_features.iloc[:len(X_test)]])
y_pred = (y_pred_prob > 0.5).astype(int)  


submission = pd.DataFrame({
    'id': df_test['id'],
    'label': y_pred.flatten()  
})

Guardar el archivo de predicciones
filename = f"biLSTM_submission_v2_{datetime.now().strftime('%Y%m%d_%H%M')}.csv"
submission.to_csv(filename, columns=['id', 'label'], index=False)

Mensaje final
print(f" Submission generada correctamente: '{filename}'")
print(submission.head())


c:\Users\anaeg\Desktop\LBBYs_CH2\spacy_env\Lib\site-packages\keras\src\layers\core\embedding.py:90: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


Model: "functional_8"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer_15      │ (None, 100)       │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ embedding_8         │ (None, 100, 100)  │  2,000,000 │ input_layer_15[0… │
│ (Embedding)         │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ bidirectional_8     │ (None, 100, 512)  │    731,136 │ embedding_8[0][0] │
│ (Bidirectional)     │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_7 (Dropout) │ (None, 100, 512)  │          0 │ bidirectional_8[… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ attention_8         │ (None, 100, 512)  │          0 │ dropout_7[0][0],  │
│ (Attention)         │                   │            │ dropout_7[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ global_average_poo… │ (None, 512)       │          0 │ attention_8[0][0] │
│ (GlobalAveragePool… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ input_layer_16      │ (None, 139)       │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ concatenate_7       │ (None, 651)       │          0 │ global_average_p… │
│ (Concatenate)       │                   │            │ input_layer_16[0… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_8 (Dense)     │ (None, 1)         │        652 │ concatenate_7[0]… │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 2,731,788 (10.42 MB)

 Trainable params: 2,731,788 (10.42 MB)

 Non-trainable params: 0 (0.00 B)

Epoch 1/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 28s 114ms/step - accuracy: 0.4831 - loss: 0.6973 - val_accuracy: 0.5849 - val_loss: 0.6861 - learning_rate: 0.0010
Epoch 2/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 24s 107ms/step - accuracy: 0.5344 - loss: 0.6936 - val_accuracy: 0.4034 - val_loss: 0.7027 - learning_rate: 0.0010
Epoch 3/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 27s 122ms/step - accuracy: 0.5011 - loss: 0.6928 - val_accuracy: 0.4603 - val_loss: 0.6961 - learning_rate: 0.0010
Epoch 4/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 24s 106ms/step - accuracy: 0.5156 - loss: 0.6863 - val_accuracy: 0.5626 - val_loss: 0.6707 - learning_rate: 5.0000e-04
Epoch 5/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 23s 103ms/step - accuracy: 0.6863 - loss: 0.6173 - val_accuracy: 0.6391 - val_loss: 0.7193 - learning_rate: 5.0000e-04
120/120 ━━━━━━━━━━━━━━━━━━━━ 3s 22ms/step
✅ Submission generada correctamente: 'biLSTM_submission_v2_20250515_1128.csv'
            id  label
0  dc32e5ffa8b      1
1  aa49bb41cab      1
2  dddc8d12ac1      1
3  bcfe8f51667

### Intento 7: Tokenización OOV para cobertura de vocabulario desconocido

In [ ]:
import pandas as pd
import numpy as np
import string
import spacy
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.layers import Dense, Embedding, Bidirectional, LSTM, Input, Dropout, GlobalAveragePooling1D, Attention, Concatenate
from tensorflow.keras.models import Model
from tensorflow.keras.optimizers import Adam
from sklearn.metrics import classification_report
from sklearn.model_selection import train_test_split
from sklearn.utils.class_weight import compute_class_weight
from tensorflow.keras.callbacks import ReduceLROnPlateau
from datetime import datetime





max_vocab = 10000  
tokenizer = Tokenizer(num_words=max_vocab, oov_token="<OOV>")
tokenizer.fit_on_texts(df['processed_statement'])


X_train = tokenizer.texts_to_sequences(df['processed_statement'])
X_test = tokenizer.texts_to_sequences(df_test['processed_statement'])


MAX_SEQUENCE_LENGTH = 100  
X_train = pad_sequences(X_train, maxlen=MAX_SEQUENCE_LENGTH)
X_test = pad_sequences(X_test, maxlen=MAX_SEQUENCE_LENGTH)


subject_encoded = pd.get_dummies(df['processed_subject'])
speaker_encoded = pd.get_dummies(df['speaker_job'])
party_affiliation_encoded = pd.get_dummies(df['party_affiliation_uni'])


X_additional_features = pd.concat([subject_encoded, speaker_encoded, party_affiliation_encoded], axis=1)


y = df['label'].values


X_train, X_val, y_train, y_val = train_test_split(X_train, y, test_size=0.2, random_state=42)


class_weights = compute_class_weight('balanced', classes=np.unique(y_train), y=y_train)


input_text = Input(shape=(MAX_SEQUENCE_LENGTH,))
input_additional = Input(shape=(X_additional_features.shape[1],))  


embedding_layer = Embedding(input_dim=max_vocab, output_dim=100, input_length=MAX_SEQUENCE_LENGTH)(input_text)


lstm_layer = Bidirectional(LSTM(256, return_sequences=True))(embedding_layer)


dropout_layer = Dropout(0.5)(lstm_layer)


attention_layer = Attention()([dropout_layer, dropout_layer])


pooling_layer = GlobalAveragePooling1D()(attention_layer)


merged = Concatenate()([pooling_layer, input_additional])


output_layer = Dense(1, activation='sigmoid')(merged)


model = Model(inputs=[input_text, input_additional], outputs=output_layer)


reduce_lr = ReduceLROnPlateau(monitor='val_loss', factor=0.5, patience=2, min_lr=1e-6)


model.compile(optimizer=Adam(), loss='binary_crossentropy', metrics=['accuracy'])


model.summary()


model.fit([X_train, X_additional_features.iloc[:len(X_train)]], y_train, epochs=5, batch_size=32,
          validation_data=([X_val, X_additional_features.iloc[:len(X_val)]], y_val),
          class_weight={0: class_weights[0], 1: class_weights[1]},
          callbacks=[reduce_lr])


y_pred_prob = model.predict([X_test, X_additional_features.iloc[:len(X_test)]])
y_pred = (y_pred_prob > 0.5).astype(int)  


submission = pd.DataFrame({
    'id': df_test['id'],
    'label': y_pred.flatten()  
})


filename = f"biLSTM_submission_v2_{datetime.now().strftime('%Y%m%d_%H%M')}.csv"
submission.to_csv(filename, columns=['id', 'label'], index=False)


print(f" Submission generada correctamente: '{filename}'")
print(submission.head())


c:\Users\anaeg\Desktop\LBBYs_CH2\spacy_env\Lib\site-packages\keras\src\layers\core\embedding.py:90: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


Model: "functional_9"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer_17      │ (None, 100)       │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ embedding_9         │ (None, 100, 100)  │  1,000,000 │ input_layer_17[0… │
│ (Embedding)         │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ bidirectional_9     │ (None, 100, 512)  │    731,136 │ embedding_9[0][0] │
│ (Bidirectional)     │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_8 (Dropout) │ (None, 100, 512)  │          0 │ bidirectional_9[… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ attention_9         │ (None, 100, 512)  │          0 │ dropout_8[0][0],  │
│ (Attention)         │                   │            │ dropout_8[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ global_average_poo… │ (None, 512)       │          0 │ attention_9[0][0] │
│ (GlobalAveragePool… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ input_layer_18      │ (None, 139)       │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ concatenate_8       │ (None, 651)       │          0 │ global_average_p… │
│ (Concatenate)       │                   │            │ input_layer_18[0… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_9 (Dense)     │ (None, 1)         │        652 │ concatenate_8[0]… │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 1,731,788 (6.61 MB)

 Trainable params: 1,731,788 (6.61 MB)

 Non-trainable params: 0 (0.00 B)

Epoch 1/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 60s 255ms/step - accuracy: 0.4834 - loss: 0.6970 - val_accuracy: 0.5358 - val_loss: 0.6903 - learning_rate: 0.0010
Epoch 2/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 55s 246ms/step - accuracy: 0.5227 - loss: 0.6935 - val_accuracy: 0.5497 - val_loss: 0.6900 - learning_rate: 0.0010
Epoch 3/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 55s 245ms/step - accuracy: 0.5621 - loss: 0.6908 - val_accuracy: 0.5564 - val_loss: 0.6899 - learning_rate: 0.0010
Epoch 4/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 56s 249ms/step - accuracy: 0.5483 - loss: 0.6880 - val_accuracy: 0.6190 - val_loss: 0.6672 - learning_rate: 0.0010
Epoch 5/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 54s 241ms/step - accuracy: 0.6892 - loss: 0.6148 - val_accuracy: 0.6095 - val_loss: 0.7042 - learning_rate: 0.0010
120/120 ━━━━━━━━━━━━━━━━━━━━ 7s 53ms/step
✅ Submission generada correctamente: 'biLSTM_submission_v2_20250515_1133.csv'
            id  label
0  dc32e5ffa8b      0
1  aa49bb41cab      0
2  dddc8d12ac1      1
3  bcfe8f51667      1


### Intento 8: Bigramas y bloque de Multi-Head Attention

In [ ]:
import pandas as pd
import numpy as np
import string
import spacy
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.layers import Dense, Embedding, Bidirectional, LSTM, Input, Dropout, GlobalAveragePooling1D, Attention, Concatenate, MultiHeadAttention
from tensorflow.keras.models import Model
from tensorflow.keras.optimizers import Nadam
from sklearn.metrics import classification_report
from sklearn.model_selection import train_test_split
from sklearn.utils.class_weight import compute_class_weight
from tensorflow.keras.callbacks import ReduceLROnPlateau, LearningRateScheduler
from tensorflow.keras.regularizers import l2
from datetime import datetime


def get_ngrams(text, n=2):
    tokens = text.split()  
    ngrams = [' '.join(tokens[i:i+n]) for i in range(len(tokens)-n+1)]  
    return ' '.join(ngrams) 


df['processed_statement'] = df['processed_statement'].apply(lambda x: get_ngrams(x, n=2)) 
df_test['processed_statement'] = df_test['processed_statement'].apply(lambda x: get_ngrams(x, n=2)) 

max_vocab = 20000  
tokenizer = Tokenizer(num_words=max_vocab, oov_token="<OOV>")
tokenizer.fit_on_texts(df['processed_statement'])  

X_train = tokenizer.texts_to_sequences(df['processed_statement'])
X_test = tokenizer.texts_to_sequences(df_test['processed_statement'])


MAX_SEQUENCE_LENGTH = 100
X_train = pad_sequences(X_train, maxlen=MAX_SEQUENCE_LENGTH)
X_test = pad_sequences(X_test, maxlen=MAX_SEQUENCE_LENGTH)


subject_encoded = pd.get_dummies(df['processed_subject'])
speaker_encoded = pd.get_dummies(df['speaker'])
party_affiliation_encoded = pd.get_dummies(df['party_affiliation'])


X_additional_features = pd.concat([subject_encoded, speaker_encoded, party_affiliation_encoded], axis=1)
y = df['label'].values
X_train, X_val, y_train, y_val = train_test_split(X_train, y, test_size=0.2, random_state=42)
class_weights = compute_class_weight('balanced', classes=np.unique(y_train), y=y_train)

input_text = Input(shape=(MAX_SEQUENCE_LENGTH,))
input_additional = Input(shape=(X_additional_features.shape[1],))  

embedding_layer = Embedding(input_dim=max_vocab, output_dim=100, input_length=MAX_SEQUENCE_LENGTH)(input_text)

lstm_layer = Bidirectional(LSTM(512, return_sequences=True))(embedding_layer)
dropout_layer = Dropout(0.5)(lstm_layer)

attention_layer = MultiHeadAttention(num_heads=8, key_dim=256)(dropout_layer, dropout_layer)

pooling_layer = GlobalAveragePooling1D()(attention_layer)

dense_layer = Dense(128, activation='relu', kernel_regularizer=l2(0.001))(pooling_layer)

merged = Concatenate()([dense_layer, input_additional])

output_layer = Dense(1, activation='sigmoid')(merged)

model = Model(inputs=[input_text, input_additional], outputs=output_layer)

reduce_lr = ReduceLROnPlateau(monitor='val_loss', factor=0.5, patience=2, min_lr=1e-6)

def lr_schedule(epoch):
    if epoch < 10:
        return 1e-3
    elif epoch < 20:
        return 1e-4
    else:
        return 1e-5

lr_scheduler = LearningRateScheduler(lr_schedule)

model.compile(optimizer=Nadam(), loss='binary_crossentropy', metrics=['accuracy'])


model.summary()


model.fit([X_train, X_additional_features.iloc[:len(X_train)]], y_train, epochs=5, batch_size=32,
          validation_data=([X_val, X_additional_features.iloc[:len(X_val)]], y_val),
          class_weight={0: class_weights[0], 1: class_weights[1]},
          callbacks=[reduce_lr, lr_scheduler])


y_pred_prob = model.predict([X_test, X_additional_features.iloc[:len(X_test)]])
y_pred = (y_pred_prob > 0.5).astype(int)  


submission = pd.DataFrame({
    'id': df_test['id'],
    'label': y_pred.flatten()  
})


filename = f"biLSTM_submission_v2_{datetime.now().strftime('%Y%m%d_%H%M')}.csv"
submission.to_csv(filename, columns=['id', 'label'], index=False)


print(f" Submission generada correctamente: '{filename}'")
print(submission.head())

c:\Users\anaeg\Desktop\LBBYs_CH2\spacy_env\Lib\site-packages\keras\src\layers\core\embedding.py:90: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


Model: "functional_10"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer_19      │ (None, 100)       │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ embedding_10        │ (None, 100, 100)  │  2,000,000 │ input_layer_19[0… │
│ (Embedding)         │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ bidirectional_10    │ (None, 100, 1024) │  2,510,848 │ embedding_10[0][… │
│ (Bidirectional)     │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_9 (Dropout) │ (None, 100, 1024) │          0 │ bidirectional_10… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ multi_head_attenti… │ (None, 100, 1024) │  8,395,776 │ dropout_9[0][0],  │
│ (MultiHeadAttentio… │                   │            │ dropout_9[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ global_average_poo… │ (None, 1024)      │          0 │ multi_head_atten… │
│ (GlobalAveragePool… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_10 (Dense)    │ (None, 128)       │    131,200 │ global_average_p… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ input_layer_20      │ (None, 2746)      │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ concatenate_9       │ (None, 2874)      │          0 │ dense_10[0][0],   │
│ (Concatenate)       │                   │            │ input_layer_20[0… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_11 (Dense)    │ (None, 1)         │      2,875 │ concatenate_9[0]… │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 13,040,699 (49.75 MB)

 Trainable params: 13,040,699 (49.75 MB)

 Non-trainable params: 0 (0.00 B)

Epoch 1/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 180s 786ms/step - accuracy: 0.4427 - loss: 0.7679 - val_accuracy: 0.4939 - val_loss: 0.6931 - learning_rate: 0.0010
Epoch 2/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 175s 779ms/step - accuracy: 0.5684 - loss: 0.6896 - val_accuracy: 0.5123 - val_loss: 0.6919 - learning_rate: 0.0010
Epoch 3/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 195s 869ms/step - accuracy: 0.6142 - loss: 0.6859 - val_accuracy: 0.5084 - val_loss: 0.6928 - learning_rate: 0.0010
Epoch 4/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 189s 844ms/step - accuracy: 0.6270 - loss: 0.6863 - val_accuracy: 0.5173 - val_loss: 0.6921 - learning_rate: 5.0000e-04
Epoch 5/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 181s 810ms/step - accuracy: 0.6617 - loss: 0.6803 - val_accuracy: 0.5112 - val_loss: 0.6924 - learning_rate: 0.0010
120/120 ━━━━━━━━━━━━━━━━━━━━ 22s 180ms/step
✅ Submission generada correctamente: 'biLSTM_submission:v2_20250515_1148.csv'
            id  label
0  dc32e5ffa8b      0
1  aa49bb41cab      1
2  dddc8d12ac1      1
3  bcfe8f51

In [25]:
import os
print(os.path.exists('glove.6B.100d.txt'))



True


### Intento 9: Embeddings GloVe preentrenados

In [ ]:
import pandas as pd
import numpy as np
import string
import spacy
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.layers import Dense, Embedding, Bidirectional, LSTM, Dropout, GlobalAveragePooling1D
from tensorflow.keras.models import Model
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.callbacks import ReduceLROnPlateau
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.utils.class_weight import compute_class_weight
from tensorflow.keras.regularizers import l2
import tensorflow as tf


max_vocab = 20000  
tokenizer = Tokenizer(num_words=max_vocab, oov_token="<OOV>")
tokenizer.fit_on_texts(df['processed_statement'])

X_train = tokenizer.texts_to_sequences(df['processed_statement'])
X_test = tokenizer.texts_to_sequences(df_test['processed_statement'])


MAX_SEQUENCE_LENGTH = 100
X_train = pad_sequences(X_train, maxlen=MAX_SEQUENCE_LENGTH)
X_test = pad_sequences(X_test, maxlen=MAX_SEQUENCE_LENGTH)

embedding_dim = 100  
embeddings_index = {}
with open('glove.6B.100d.txt', 'r', encoding='utf-8') as f:
    for line in f:
        values = line.split()
        word = values[0]
        vector = np.asarray(values[1:], dtype='float32')
        embeddings_index[word] = vector

embedding_matrix = np.zeros((max_vocab, embedding_dim))
for word, index in tokenizer.word_index.items():
    if index < max_vocab:
        embedding_vector = embeddings_index.get(word)
        if embedding_vector is not None:
            embedding_matrix[index] = embedding_vector

subject_encoded = pd.get_dummies(df['processed_subject'])
speaker_encoded = pd.get_dummies(df['speaker'])
party_affiliation_encoded = pd.get_dummies(df['party_affiliation'])
X_additional_features = pd.concat([subject_encoded, speaker_encoded, party_affiliation_encoded], axis=1)


y = df['label'].values


X_train, X_val, y_train, y_val = train_test_split(X_train, y, test_size=0.2, random_state=42)


class_weights = compute_class_weight('balanced', classes=np.unique(y_train), y=y_train)


input_text = tf.keras.Input(shape=(MAX_SEQUENCE_LENGTH,))
input_additional = tf.keras.Input(shape=(X_additional_features.shape[1],))  

embedding_layer = Embedding(input_dim=max_vocab, output_dim=embedding_dim, weights=[embedding_matrix], input_length=MAX_SEQUENCE_LENGTH, trainable=False)(input_text)

lstm_layer = Bidirectional(LSTM(256, return_sequences=True))(embedding_layer)
dropout_layer = Dropout(0.5)(lstm_layer)


pooling_layer = GlobalAveragePooling1D()(dropout_layer)

merged = tf.keras.layers.concatenate([pooling_layer, input_additional])

output_layer = Dense(1, activation='sigmoid')(merged)


model = Model(inputs=[input_text, input_additional], outputs=output_layer)


reduce_lr = ReduceLROnPlateau(monitor='val_loss', factor=0.5, patience=2, min_lr=1e-6)


model.compile(optimizer=Adam(), loss='binary_crossentropy', metrics=['accuracy'])


model.summary()

model.fit([X_train, X_additional_features.iloc[:len(X_train)]], y_train, epochs=5, batch_size=32,
          validation_data=([X_val, X_additional_features.iloc[:len(X_val)]], y_val),
          class_weight={0: class_weights[0], 1: class_weights[1]},
          callbacks=[reduce_lr])


y_pred_prob = model.predict([X_test, X_additional_features.iloc[:len(X_test)]])
y_pred = (y_pred_prob > 0.5).astype(int)  


submission = pd.DataFrame({
    'id': df_test['id'],
    'label': y_pred.flatten()  
})


filename = f"biLSTM_submission_v2_{datetime.now().strftime('%Y%m%d_%H%M')}.csv"
submission.to_csv(filename, columns=['id', 'label'], index=False)


print(f" Submission generada correctamente: '{filename}'")
print(submission.head())


c:\Users\anaeg\Desktop\LBBYs_CH2\spacy_env\Lib\site-packages\keras\src\layers\core\embedding.py:90: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


Model: "functional_11"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer_22      │ (None, 100)       │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ embedding_11        │ (None, 100, 100)  │  2,000,000 │ input_layer_22[0… │
│ (Embedding)         │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ bidirectional_11    │ (None, 100, 512)  │    731,136 │ embedding_11[0][… │
│ (Bidirectional)     │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_11          │ (None, 100, 512)  │          0 │ bidirectional_11… │
│ (Dropout)           │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ global_average_poo… │ (None, 512)       │          0 │ dropout_11[0][0]  │
│ (GlobalAveragePool… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ input_layer_23      │ (None, 2746)      │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ concatenate_10      │ (None, 3258)      │          0 │ global_average_p… │
│ (Concatenate)       │                   │            │ input_layer_23[0… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_12 (Dense)    │ (None, 1)         │      3,259 │ concatenate_10[0… │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 2,734,395 (10.43 MB)

 Trainable params: 734,395 (2.80 MB)

 Non-trainable params: 2,000,000 (7.63 MB)

Epoch 1/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 42s 177ms/step - accuracy: 0.5262 - loss: 0.6933 - val_accuracy: 0.4911 - val_loss: 0.7089 - learning_rate: 0.0010
Epoch 2/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 39s 172ms/step - accuracy: 0.5727 - loss: 0.6755 - val_accuracy: 0.6318 - val_loss: 0.6640 - learning_rate: 0.0010
Epoch 3/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 36s 162ms/step - accuracy: 0.6058 - loss: 0.6737 - val_accuracy: 0.6380 - val_loss: 0.6517 - learning_rate: 0.0010
Epoch 4/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 34s 153ms/step - accuracy: 0.6182 - loss: 0.6621 - val_accuracy: 0.5732 - val_loss: 0.6829 - learning_rate: 0.0010
Epoch 5/5
224/224 ━━━━━━━━━━━━━━━━━━━━ 34s 153ms/step - accuracy: 0.6145 - loss: 0.6556 - val_accuracy: 0.6330 - val_loss: 0.6505 - learning_rate: 0.0010
120/120 ━━━━━━━━━━━━━━━━━━━━ 5s 38ms/step
✅ Submission generada correctamente: 'biLSTM_submission_v2_20250515_1218.csv'
            id  label
0  dc32e5ffa8b      0
1  aa49bb41cab      1
2  dddc8d12ac1      1
3  bcfe8f51667      1


### Intento 10: RoBERTa fijo + BiLSTM multimodal


In [ ]:
from transformers import TFRobertaModel, RobertaTokenizer
import numpy as np
from tensorflow.keras.layers import Input, Bidirectional, LSTM, Dense, Dropout, Concatenate, GlobalAveragePooling1D
from tensorflow.keras.models import Model
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.callbacks import EarlyStopping, ReduceLROnPlateau
from sklearn.model_selection import train_test_split

tokenizer = RobertaTokenizer.from_pretrained('roberta-base')
roberta_model = TFRobertaModel.from_pretrained('roberta-base')
roberta_model.trainable = False

def tokenize_texts(texts):
    return tokenizer(texts.tolist(), padding=True, truncation=True, max_length=128, return_tensors="tf")

X_train_tokens = tokenize_texts(df['processed_statement'])
X_test_tokens = tokenize_texts(df_test['processed_statement'])

X_train = X_train_tokens['input_ids'].numpy()
X_test = X_test_tokens['input_ids'].numpy()
y = df['label'].values

X_train, X_val, y_train, y_val = train_test_split(X_train, y, test_size=0.2, random_state=42)

input_text = Input(shape=(128,), dtype='int32')
input_additional = Input(shape=(X_additional_features.shape[1],))

roberta_outputs = roberta_model(input_text)[0]

lstm_layer = Bidirectional(LSTM(256, return_sequences=True))(roberta_outputs)
dropout_layer = Dropout(0.5)(lstm_layer)
pooling_layer = GlobalAveragePooling1D()(dropout_layer)

merged = Concatenate()([pooling_layer, input_additional])
output_layer = Dense(1, activation='sigmoid')(merged)

model = Model(inputs=[input_text, input_additional], outputs=output_layer)
model.compile(optimizer=Adam(learning_rate=5e-5), loss='binary_crossentropy', metrics=['accuracy'])

early_stopping = EarlyStopping(monitor='val_loss', patience=2, restore_best_weights=True)
reduce_lr = ReduceLROnPlateau(monitor='val_loss', factor=0.5, patience=2, min_lr=1e-6)

model.fit([X_train, X_additional_features.iloc[:len(X_train)]], y_train, epochs=5, batch_size=16,
          validation_data=([X_val, X_additional_features.iloc[:len(X_val)]], y_val),
          callbacks=[early_stopping, reduce_lr])

y_pred_prob = model.predict([X_test, X_additional_features.iloc[:len(X_test)]])
y_pred = (y_pred_prob > 0.5).astype(int)

import pandas as pd
from datetime import datetime

submission = pd.DataFrame({'id': df_test['id'], 'label': y_pred.flatten()})
filename = f"roberta_bilstm_submission_v2_{datetime.now().strftime('%Y%m%d_%H%M')}.csv"
submission.to_csv(filename, columns=['id', 'label'], index=False)

print(f"Submission generada correctamente: '{filename}'")


: 

## **Conclusión**

De los diez prototipos de BiLSTM que exploramos, el Intento 3 fue el que mejor resultado nos entregó: al combinar un BiLSTM más profundo (256 unidades), aplicar un dropout fuerte, usar pesos de clase balanceados y un callback de reducción de la tasa de aprendizaje, la validación alcanzó un 67,1 % en el cuarto epoch antes de estabilizarse en torno al 61 %.